### DISCUSSION TITLE
Is it possible to compile the engine on ARM Linux ?

### DISCUSSION THREAD

**Discussion by Darknior** [Wohlstand/TheXTech#385] [Q&A]
**Describe the issue**

Is it possible to compile the engine on ARM Linux ?
On release page, there is only an Windows ARM version 64 bit.

**Example Case**

I'm a Batocera user and contributor. This is a game platform like Recallbox, EmuElec, etc ...

https://batocera.org/

We can add your engine in Batocera for PC version, users will love it for sure.
But we also working on Raspberry PI3, Odroid N2+ etc ... many boards ...
Please if it was possible to port it also on ARM linux, 64 bit, we will can add it for every one :)
Thanks a lot to read me

**Comment by Wohlstand** [comment]
This engine is already buildable for ARM:
- I do have Orange Pi PC thing where it works
- @ChristianSilvermoon has some of Raspberry Pi-like boards where the game works
- other folks at my chat has successfully built the game from sources and ran it on their ARM devices
- game builds and works on Android with ARM processors easily
- Supported both ARM32 and ARM64
- Also buildable and runnable for Windows ARM64 too, my friend actually has a phone where he installed Windows 10 for his experiments and game works successfully

**Comment by Wohlstand** [comment]
P.S. Just use the regular manual of Linux building, and in a case you performing a cross compilation, you should specify the CMake toolchain file that plugs the compatible cross-compiler and runs it.

**Comment by Darknior** [comment]
Ok thanks a lot i will try, like you d'ont share these two versions i've think it was maybe not compatible.
I will report here when it will be integrated on Batocera :)

**Comment by ChristianSilvermoon** [comment]
I have succesfully built the game from source on:
* [Pine64 PinePhone (3GB RAM/32GB eMMC variant)](https://www.pine64.org/pinephone/), running [Mobian](https://mobian-project.org/). A 64-bit ARM System running a 64-bit OS.
* [Raspberry Pi 4B (8GB RAM variant)](https://www.raspberrypi.com/products/raspberry-pi-4-model-b/), running [Raspberry Pi OS (formerly Raspbian) Buster](https://www.raspberrypi.com/software/) as initially configured ([RetroPie](https://retropie.org.uk/), some safe shutdown scripts, and some FOSS games, etc) by [ExperimentalPi](https://experimentalpi.com/) for their **PiBoy DMG** handheld. 64-bit ARM system running a 32-bit OS in this case.

The standard build instructions for "[Building on Linux or xBSD](https://github.com/Wohlstand/TheXTech/wiki/Building-on-Linux-or-xBSD)" should work just fine :P

**Comment by Darknior** [comment]
> I have succesfully built the game from source on:
> The standard build instructions for "[Building on Linux or xBSD](https://github.com/Wohlstand/TheXTech/wiki/Building-on-Linux-or-xBSD)" should work just fine :P

For the moment i don't have a pi3 or PI4 at home to compile and test, but is it possible to share here the last engine compiled for ARM 32 and 64 ? Only to try it, not the entire game.

**Comment by ChristianSilvermoon** [comment]
> > I have succesfully built the game from source on:
> > The standard build instructions for "[Building on Linux or xBSD](https://github.com/Wohlstand/TheXTech/wiki/Building-on-Linux-or-xBSD)" should work just fine :P
> 
> For the moment i don't have a pi3 or PI4 at home to compile and test, but is it possible to share here the last engine compiled for ARM 32 and 64 ? Only to try it, not the entire game.

I don't believe I built a portable binary, actually. I'd need to rebuild it with different flags.

EDIT: Attempting to compile according to
```
cmake -G Ninja -DCMAKE_BUILD_TYPE=MinSizeRel -DUSE_SYSTEM_SDL2=OFF -DUSE_SYSTEM_LIBS=OFF -DUSE_STATIC_LIBC=ON -DPGE_SHARED_SDLMIXER=OFF ..
ninja
```

**Comment by ChristianSilvermoon** [comment]
Not having any luck with the current Master or Devel branch on my Pi 4 atm :thinking: 

I attatched the point it failed... any advice @Wohlstand  ?
```
[140/140] Linking CXX executable output/bin/thextech
FAILED: output/bin/thextech
: && /usr/bin/c++  -Wall -Wextra -pedantic -Wno-variadic-macros -ffloat-store -Os -DNDEBUG  -no-pie CMakeFiles/thextech.dir/src/compat.cpp.o CMakeFiles/thextech.dir/src/game_main.cpp.o CMakeFiles/thextech.dir/src/globals.cpp.o CMakeFiles/thextech.dir/src/npc.cpp.o CMakeFiles/thextech.dir/src/sound.cpp.o CMakeFiles/thextech.dir/src/frame_timer.cpp.o CMakeFiles/thextech.dir/src/main/world_loop.cpp.o CMakeFiles/thextech.dir/src/main/world_file.cpp.o CMakeFiles/thextech.dir/src/main/game_info.cpp.o CMakeFiles/thextech.dir/src/main/game_loop.cpp.o CMakeFiles/thextech.dir/src/main/gameplay_timer.cpp.o CMakeFiles/thextech.dir/src/main/player_frames.cpp.o CMakeFiles/thextech.dir/src/main/setup_vars.cpp.o CMakeFiles/thextech.dir/src/main/setup_physics.cpp.o CMakeFiles/thextech.dir/src/main/speedrunner.cpp.o CMakeFiles/thextech.dir/src/main/record.cpp.o CMakeFiles/thextech.dir/src/main/game_save.cpp.o CMakeFiles/thextech.dir/src/main/main_config.cpp.o CMakeFiles/thextech.dir/src/main/level_file.cpp.o CMakeFiles/thextech.dir/src/main/menu_loop.cpp.o CMakeFiles/thextech.dir/src/main/menu_main.cpp.o CMakeFiles/thextech.dir/src/main/outro_loop.cpp.o CMakeFiles/thextech.dir/src/main/trees.cpp.o CMakeFiles/thextech.dir/src/main/QuadTree/LooseQuadtree-impl.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_update2.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_update.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_background.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_print.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_message.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_editor.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_credits.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_hud.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_enter_screen.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_draw_player.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_special_frames.cpp.o CMakeFiles/thextech.dir/src/graphics/gfx_screen.cpp.o CMakeFiles/thextech.dir/src/change_res.cpp.o CMakeFiles/thextech.dir/src/effect.cpp.o CMakeFiles/thextech.dir/src/collision.cpp.o CMakeFiles/thextech.dir/src/load_gfx.cpp.o CMakeFiles/thextech.dir/src/layers.cpp.o CMakeFiles/thextech.dir/src/custom.cpp.o CMakeFiles/thextech.dir/src/graphics.cpp.o CMakeFiles/thextech.dir/src/editor/editor.cpp.o CMakeFiles/thextech.dir/src/editor/new_editor.cpp.o CMakeFiles/thextech.dir/src/editor/write_common.cpp.o CMakeFiles/thextech.dir/src/editor/write_level.cpp.o CMakeFiles/thextech.dir/src/editor/write_world.cpp.o CMakeFiles/thextech.dir/src/main.cpp.o CMakeFiles/thextech.dir/src/blocks.cpp.o CMakeFiles/thextech.dir/src/gfx.cpp.o CMakeFiles/thextech.dir/src/player.cpp.o CMakeFiles/thextech.dir/src/rand.cpp.o CMakeFiles/thextech.dir/src/sorting.cpp.o CMakeFiles/thextech.dir/src/npc/npc_hit.cpp.o CMakeFiles/thextech.dir/src/npc/npc_kill.cpp.o CMakeFiles/thextech.dir/src/npc/npc_update.cpp.o CMakeFiles/thextech.dir/src/npc/npc_frames.cpp.o CMakeFiles/thextech.dir/src/npc/npc_bonus.cpp.o CMakeFiles/thextech.dir/src/player/player_update.cpp.o CMakeFiles/thextech.dir/src/control/joystick.cpp.o CMakeFiles/thextech.dir/src/frm_main.cpp.o CMakeFiles/thextech.dir/src/main/cheat_code.cpp.o CMakeFiles/thextech.dir/src/editor/text_input.cpp.o CMakeFiles/thextech.dir/src/mixer_platform.cpp.o CMakeFiles/thextech.dir/lib/util.cpp.o CMakeFiles/thextech.dir/lib/AppPath/app_path.cpp.o CMakeFiles/thextech.dir/lib/DirManager/dirman.cpp.o CMakeFiles/thextech.dir/lib/DirManager/dirman_posix.cpp.o CMakeFiles/thextech.dir/lib/fmt/fmt_format.cpp.o CMakeFiles/thextech.dir/lib/fmt/fmt_ostream.cpp.o CMakeFiles/thextech.dir/lib/fmt/fmt_posix.cpp.o CMakeFiles/thextech.dir/lib/fmt/fmt_printf.cpp.o CMakeFiles/thextech.dir/lib/IniProcessor/ini_processing.cpp.o CMakeFiles/thextech.dir/lib/Logger/private/logger.cpp.o CMakeFiles/thextech.dir/lib/Logger/private/logger_desktop.cpp.o CMakeFiles/thextech.dir/lib/Utils/maths.cpp.o CMakeFiles/thextech.dir/lib/Utils/files.cpp.o CMakeFiles/thextech.dir/lib/Utils/strings.cpp.o CMakeFiles/thextech.dir/lib/Utils/elapsed_timer.cpp.o CMakeFiles/thextech.dir/lib/Utils/dir_list_ci.cpp.o CMakeFiles/thextech.dir/lib/FileMapper/file_mapper.cpp.o CMakeFiles/thextech.dir/3rdparty/tinyfiledialogs/tinyfiledialogs.c.o CMakeFiles/thextech.dir/lib/CrashHandler/crash_handler.cpp.o CMakeFiles/thextech.dir/lib/Graphics/graphics_funcs.cpp.o CMakeFiles/thextech.dir/lib/Graphics/image_size.cpp.o CMakeFiles/thextech.dir/lib/Graphics/size.cpp.o CMakeFiles/thextech.dir/lib/Graphics/sizef.cpp.o CMakeFiles/thextech.dir/lib/InterProcess/editor_pipe.cpp.o CMakeFiles/thextech.dir/lib/InterProcess/intproc.cpp.o  -o output/bin/thextech  -Wl,-rpath,:  lib/libpgefl.a  output/lib/libFreeImageLite.a  output/lib/libSDL2_mixer_ext.a  output/lib/libFLAC.a  output/lib/libfluidlite.a  output/lib/libopusfile.a  output/lib/libopus.a  output/lib/libogg.a  output/lib/libADLMIDI.a  output/lib/libOPNMIDI.a  output/lib/libtimidity_sdl2.a  output/lib/libgme.a  output/lib/libxmp.a  output/lib/libmodplug.a  output/lib/libzlib.a  output/lib/libSDL2.a  /usr/lib/arm-linux-gnueabihf/libGL.so  /usr/lib/arm-linux-gnueabihf/libpthread.so  /usr/lib/arm-linux-gnueabihf/libdl.so && :
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpivideo.c.o): in function `AddDispManXDisplay':
SDL_rpivideo.c:(.text+0x278): undefined reference to `vc_dispmanx_display_open'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x288): undefined reference to `vc_dispmanx_display_get_info'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x298): undefined reference to `vc_dispmanx_display_close'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x2d0): undefined reference to `vc_tv_get_display_state'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x2ec): undefined reference to `vc_tv_hdmi_get_property'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpivideo.c.o): in function `RPI_VideoInit':
SDL_rpivideo.c:(.text+0x3a0): undefined reference to `bcm_host_init'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpivideo.c.o): in function `RPI_CreateWindow':
SDL_rpivideo.c:(.text+0x49c): undefined reference to `vc_dispmanx_update_start'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x4d4): undefined reference to `vc_dispmanx_element_add'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x4f0): undefined reference to `vc_dispmanx_update_submit_sync'
/usr/bin/ld: SDL_rpivideo.c:(.text+0x580): undefined reference to `vc_dispmanx_vsync_callback'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpivideo.c.o): in function `RPI_DestroyWindow':
SDL_rpivideo.c:(.text+0x62c): undefined reference to `vc_dispmanx_vsync_callback'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpimouse.c.o): in function `RPI_WarpMouseGlobal':
SDL_rpimouse.c:(.text+0x70): undefined reference to `vc_dispmanx_update_start'
/usr/bin/ld: SDL_rpimouse.c:(.text+0xe4): undefined reference to `vc_dispmanx_element_change_attributes'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x108): undefined reference to `vc_dispmanx_update_submit'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpimouse.c.o): in function `RPI_MoveCursor':
SDL_rpimouse.c:(.text+0x17c): undefined reference to `vc_dispmanx_update_start'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x1f0): undefined reference to `vc_dispmanx_element_change_attributes'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x218): undefined reference to `vc_dispmanx_update_submit'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpimouse.c.o): in function `RPI_ShowCursor':
SDL_rpimouse.c:(.text+0x2a8): undefined reference to `vc_dispmanx_update_start'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x2b4): undefined reference to `vc_dispmanx_element_remove'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x2bc): undefined reference to `vc_dispmanx_update_submit_sync'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x340): undefined reference to `vc_dispmanx_rect_set'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x370): undefined reference to `vc_dispmanx_rect_set'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x378): undefined reference to `vc_dispmanx_update_start'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x3d0): undefined reference to `vc_dispmanx_element_add'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x3dc): undefined reference to `vc_dispmanx_update_submit_sync'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpimouse.c.o): in function `RPI_CreateCursor':
SDL_rpimouse.c:(.text+0x48c): undefined reference to `vc_dispmanx_resource_create'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x4ac): undefined reference to `vc_dispmanx_rect_set'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x4c0): undefined reference to `vc_dispmanx_resource_write_data'
/usr/bin/ld: output/lib/libSDL2.a(SDL_rpimouse.c.o): in function `RPI_FreeCursor':
SDL_rpimouse.c:(.text+0x4f4): undefined reference to `vc_dispmanx_update_start'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x500): undefined reference to `vc_dispmanx_element_remove'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x508): undefined reference to `vc_dispmanx_update_submit_sync'
/usr/bin/ld: SDL_rpimouse.c:(.text+0x518): undefined reference to `vc_dispmanx_resource_delete'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by Wohlstand** [comment]
Looks like there is a Raspberry PI-specific system library needed to be linked... :thinking:

But, try a simpler way, and use the system SDL2 flag at CMake build to use the SDL2 from the system. Otherwise, I don't know which library is needed to link.

**Comment by Wohlstand** [comment]
Okay, you need to try linking the `bcm_host` library (appears at /opt/vc/lib)

**Comment by ChristianSilvermoon** [comment]
Attempting with `-DUSE_SYSTEM_SDL2=ON` produced a functioning binary ^^

Here we go, built on 32-bit Raspbian Buster on Raspberry Pi 4B:
[thextech-armhf-devel-52a6f493.bin.tar.gz](https://github.com/Wohlstand/TheXTech/files/8380266/thextech-armhf-devel-52a6f493.bin.tar.gz)

**Comment by Darknior** [comment]
> Attempting with `-DUSE_SYSTEM_SDL2=ON` produced a functioning binary ^^
> 
> Here we go, built on 32-bit Raspbian Buster on Raspberry Pi 4B: [thextech-armhf-devel-52a6f493.bin.tar.gz](https://github.com/Wohlstand/TheXTech/files/8380266/thextech-armhf-devel-52a6f493.bin.tar.gz)

Thanks a lot for your effort :)
But Batocera is full 64 bit on PI from version 33. Now we are working on v34.
64 bit open some new doors for some emulators ...
Will it be possible to have a 64 bit version to try ?

**Comment by ChristianSilvermoon** [comment]
> > Attempting with `-DUSE_SYSTEM_SDL2=ON` produced a functioning binary ^^
> > Here we go, built on 32-bit Raspbian Buster on Raspberry Pi 4B: [thextech-armhf-devel-52a6f493.bin.tar.gz](https://github.com/Wohlstand/TheXTech/files/8380266/thextech-armhf-devel-52a6f493.bin.tar.gz)
> 
> Thanks a lot for your effort :) But Batocera is full 64 bit on PI from version 33. Now we are working on v34. 64 bit open some new doors for some emulators ... Will it be possible to have a 64 bit version to try ?

Unfortunately not from me, sorry '^^
I don't have many SD cards and don't really want to risk borking stuff. The PiBoy DMG requires some special scripts and configuration to function correctly and I'm not sure a 64-bit image is available/supported.

If you have access to a Batocera system you could perhaps try building the game directly on it. That's how I built it for my Pi was by cloning the repo onto the Pi itself and compiling it there. ^^

**Comment by Wohlstand** [comment]
Maybe try to cross-compile the stuff? :thinking:

**Comment by Wohlstand** [comment]
Btw, with a cross-compiling toolchain will be possible to run CI builds to produce Linux ARM builds automatically and deploy with releases.

**Comment by Wohlstand** [comment]
I think I'll move this into discussions.
